File: backend/randomizer.py
import random
import string
import uuid
from datetime import datetime, timedelta

def rand_num(digits=4):
    return ''.join(random.choices(string.digits, k=int(digits)))

def rand_choice(choices):
    arr = choices if isinstance(choices, list) else [s.strip() for s in str(choices).split(',')]
    return random.choice(arr)

def rand_guid():
    return str(uuid.uuid4())

def rand_date(start, end):
    fmt = "%Y-%m-%d"
    dt_start = datetime.strptime(start, fmt)
    dt_end = datetime.strptime(end, fmt)
    days = (dt_end - dt_start).days
    rand_days = random.randint(0, days)
    return (dt_start + timedelta(days=rand_days)).strftime(fmt)

def rand_email():
    user = 'user' + rand_num(5)
    domains = ['example.com', 'testmail.com', 'mailinator.com']
    return f"{user}@{rand_choice(domains)}"

def apply_random_format(fmt):
    # Example: "Demo_{randNum:6}_Account", "{randChoice:a,b,c}", "{randGuid}"
    val = fmt
    for _ in range(5):  # allow nested randoms
        val = val.replace('{randGuid}', rand_guid())
        val = val.replace('{randEmail}', rand_email())
        # randNum
        val = re_sub(r'\{randNum(?::(\d+))?\}', lambda m: rand_num(m.group(1) or 4), val)
        # randChoice
        val = re_sub(r'\{randChoice:([^}]+)\}', lambda m: rand_choice(m.group(1).split(',')), val)
        # randDate
        val = re_sub(r'\{randDate:([^,}]+),([^}]+)\}', lambda m: rand_date(m.group(1), m.group(2)), val)
    return val

import re
def re_sub(pattern, repl, string):
    return re.sub(pattern, repl, string)

File: backend/utils/randomizer.js
const { v4: uuidv4 } = require('uuid');

function randNum(digits = 4) {
    const min = Math.pow(10, digits - 1);
    const max = Math.pow(10, digits) - 1;
    return Math.floor(Math.random() * (max - min + 1)) + min;
}

function randChoice(list) {
    if (!Array.isArray(list)) list = String(list).split(',');
    const trimmed = list.map(s => s.trim()).filter(Boolean);
    return trimmed[Math.floor(Math.random() * trimmed.length)];
}

function randGuid() {
    return uuidv4();
}

function randDate(start, end) {
    const startDt = new Date(start);
    const endDt = new Date(end);
    const dt = new Date(startDt.getTime() + Math.random() * (endDt.getTime() - startDt.getTime()));
    return dt.toISOString().split('T')[0];
}

function randEmail() {
    const user = `user${randNum(5)}`;
    const doms = ['example.com', 'testmail.com', 'mailinator.com'];
    return `${user}@${randChoice(doms)}`;
}

function applyRandomFormat(format) {
    // Example: "Demo_{randNum:6}_Account", "{randChoice:a,b,c}", "{randGuid}"
    return format.replace(/\{randNum(?::(\d+))?\}/g, (_, d) => randNum(d ? parseInt(d) : 4))
        .replace(/\{randChoice:([^}]+)\}/g, (_, choices) => randChoice(choices.split(',')))
        .replace(/\{randGuid\}/g, () => randGuid())
        .replace(/\{randDate:([^,}]+),([^}]+)\}/g, (_, start, end) => randDate(start, end))
        .replace(/\{randEmail\}/g, () => randEmail());
}

module.exports = { randNum, randChoice, randGuid, randDate, randEmail, applyRandomFormat };
